### PR TITLE
optimize get_head_indices with tensor operations

### PR DESCRIPTION
**Pull request by pzhanggit**
Tested on Summit with profiling in https://github.com/ORNL/HydraGNN/pull/54, about 15x speedup is observed

### CODE REVIEW

File: hydragnn/train/train_validate_test.py [hunk 1]
```diff
@@ -160,14 +160,25 @@ def get_head_indices(model, data):
     y_loc = data.y_loc
     # head size for each sample
     total_size = y_loc[:, -1]
-    head_index = []
+    # feature index for all heads
+    head_index = [None] * model.num_heads
+    # intermediate work list
+    _head_ind = [None] * batch_size
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:166]: Can we rename this variable to something more clear? Even just something saying it's temporary

File: hydragnn/train/train_validate_test.py [hunk 2]
```diff
@@ -160,14 +160,25 @@ def get_head_indices(model, data):
...
+    # track the start loc of each sample
+    sample_start = torch.cumsum(total_size, dim=0) - total_size
+    sample_start = sample_start.view(-1, 1)
+    # shape (batch_size, model.num_heads), start and end of each head for each sample
+    start_track = sample_start + y_loc[:, :-1]
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:171]: I don't understand "track". `start_index`?

File: hydragnn/train/train_validate_test.py [hunk 3]
```diff
@@ -160,14 +160,25 @@ def get_head_indices(model, data):
...
+    end_track = sample_start + y_loc[:, 1:]
+
+    # a large index tensor pool for all element in data.y
+    index_pool = torch.arange(0, end_track[-1, -1])
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:175]: `index_range`?

**Review by streeve — CHANGES_REQUESTED** [review]
It will be great to have this merged! Just requesting name changes, as usual

**Comment by jychoi-hpc** [comment]
I am just wondering, if we make `head_index` as a torch array on GPU, will it be better? 

It can get some speed up when it used with other tensor data on GPUs, e.g.,:
```
loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
```
`pred`, `data.y` are all on GPU, right?

**Comment by pzhanggit** [comment]
> I am just wondering, if we make `head_index` as a torch array on GPU, will it be better?
> 
> It can get some speed up when it used with other tensor data on GPUs, e.g.,:
> 
> ```
> loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
> ```
> 
> `pred`, `data.y` are all on GPU, right?

Good point. Currently it is a list of tenors. `head_index=[tensor_index_head0, tensor_index_head1,...]`.  The thing is that the length of `head_index` for each head is different, for example 32 for `tensor_index_head0` and 32*num_nodes for `tensor_index_head1` when batch_size=32. Not sure how to define a tensor for this, maybe padding? What do you think?
I will make the `tensor_index_head*` on GPU.

**Comment by jychoi-hpc** [comment]
So, the each element in `head_index` in a tensor (on GPU, if GPU is available). Then, it should be ok with `list`.

**Comment by pzhanggit** [comment]
> So, the each element in `head_index` in a tensor (on GPU, if GPU is available). Then, it should be ok with `list`.

I will make it so explicitly in a new commit.

File: hydragnn/train/train_validate_test.py [hunk 4]
```diff
@@ -160,14 +160,25 @@ def get_head_indices(model, data):
     y_loc = data.y_loc
     # head size for each sample
     total_size = y_loc[:, -1]
-    head_index = []
+    # feature index for all heads
+    head_index = [None] * model.num_heads
+    # intermediate work list
+    _head_ind = [None] * batch_size
+    # track the start loc of each sample
+    sample_start = torch.cumsum(total_size, dim=0) - total_size
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:168]: Along with what @jychoi-hpc was saying, I think you can specify the device (same for `arange`):
`sample_start = torch.cumsum(total_size, dim=0, out=torch.Tensor(device=device))`

Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:168]: Yeah, did so. Waiting for nodes on Summit for testing.

Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:184]: @streeve  @jychoi-hpc  The tensors should be in the same device with data.x and data.y now.

**Review by jychoi-hpc — APPROVED** [review]
It looks ok to me.

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 5]
```diff
@@ -128,7 +128,9 @@ def __update_predicted_values(self, type: list, index: list, data: Data):
             A Data object representing a structure that has atoms.
         """
         output_feature = []
-        data.y_loc = torch.zeros(1, len(type) + 1, dtype=torch.int64)
+        data.y_loc = torch.zeros(
+            1, len(type) + 1, dtype=torch.int64, device=data.y.device
```
Comment (streeve) [inline hydragnn/preprocess/serialized_dataset_loader.py:132]: This is a very nice way to reduce calls to `get_device` and simplify

File: hydragnn/train/train_validate_test.py (hunk 4 above)
Comment (streeve) [reply hydragnn/train/train_validate_test.py:184]: Ok, since you moved `y_loc` to the device, this will default there as well?

**Review by streeve — APPROVED** [review]
Looks great!

File: hydragnn/train/train_validate_test.py (hunk 4 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py:184]: Yes.
